core/src/main/java/com/linecorp/armeria/internal/Http1ObjectEncoder.java

```diff
@@ -277,26 +309,67 @@ private HttpObject convertClientHeaders(int streamId, HttpHeaders headers, boole
 
     private void convert(int streamId, HttpHeaders inHeaders,
                          io.netty.handler.codec.http.HttpHeaders outHeaders, boolean trailer,
-                         boolean isRequest) throws Http2Exception {
+                         boolean isRequest, boolean endStream,
+                         HttpHeaders additionalHeaders, HttpHeaders additionalTrailers,
+                         boolean enableServerHeader, boolean enableDateHeader,
+                         @Nullable String authorityHeader, @Nullable String userAgentHeader)
+            throws Http2Exception {
 
         ArmeriaHttpUtil.toNettyHttp1(
-                streamId, inHeaders, outHeaders, HttpVersion.HTTP_1_1, trailer, isRequest);
+                streamId, inHeaders, additionalHeaders, additionalTrailers, outHeaders,
+                HttpVersion.HTTP_1_1, trailer, isRequest, endStream);
 
         outHeaders.remove(ExtensionHeaderNames.STREAM_ID.text());
+
         if (server) {
             outHeaders.remove(ExtensionHeaderNames.SCHEME.text());
+
+            if (!trailer) {
+                if (outHeaders.contains(HttpHeaderNames.CONTENT_LENGTH) &&
+                    !additionalTrailers.isEmpty()) {
+                    // We don't apply chunked encoding when the content-length header is set, which would
+                    // prevent the trailers from being sent so we go ahead and remove content-length to
+                    // force chunked encoding.
+                    outHeaders.remove(HttpHeaderNames.CONTENT_LENGTH);
+                }
+
+                if (enableServerHeader && !outHeaders.contains(HttpHeaderNames.SERVER)) {
+                    outHeaders.add(HttpHeaderNames.SERVER, ArmeriaHttpUtil.SERVER_HEADER);
+                }
+
+                if (enableDateHeader && !outHeaders.contains(HttpHeaderNames.DATE)) {
+                    outHeaders.add(HttpHeaderNames.DATE, HttpTimestampSupplier.currentTime());
+                }
+            }
         } else {
             outHeaders.remove(ExtensionHeaderNames.PATH.text());
+
+            if (!trailer) {
+                if (userAgentHeader != null && !outHeaders.contains(HttpHeaderNames.USER_AGENT)) {
+                    outHeaders.add(HttpHeaderNames.USER_AGENT, userAgentHeader);
+                }
+
+                if (authorityHeader != null && !outHeaders.contains(HttpHeaderNames.HOST)) {
+                    outHeaders.add(HttpHeaderNames.HOST, authorityHeader);
+                }
```

Comment:
Could we calculate `userAgentHeader` and `authorityHeader` until here (or `Http2ObjectEncoder`)? By doing so, we could remove `userAgentHeader` and `authorityHeader` parameter for many methods.

Reply:
Removed `userAgentHeader` and `authorityHeader`